File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        d = {'}',')',']'}
        di = {'}':'{',')':'(',']':'['}
        for i in s:
            if i in d and stk and stk[-1]==di[i]:
                stk.pop()
            else:
                stk.append(i)
        return len(stk)==0

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_set<char> open = {'(','{','['};
        unordered_map<char,char> dict = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };
        stack<char> st;
        char x;
        for (char c:s){
            if (open.find(c) != open.end()){
                st.push(c);
            }
            else{
                if (st.empty()) return false;
                x = st.top();
                st.pop();
                if (dict[c] != x){
                    return false;
                }
            }
        }
        return st.empty();
    }
};